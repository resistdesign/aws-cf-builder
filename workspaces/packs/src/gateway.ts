import { createResourcePack, SimpleCFT } from '@aws-cf-builder/utils';

export const DEFAULT_AUTH_TYPE = 'COGNITO_USER_POOLS';

export type AddGatewayConfig = {
  id: string;
  hostedZoneId: any;
  domainName: any;
  certificateArn: any;
  cloudFunction: { id: string; region?: string };
  stageName?: any;
  authorizer?: {
    providerARNs?: string[];
    scopes?: string[];
    type?: 'TOKEN' | 'COGNITO_USER_POOLS' | 'REQUEST';
    identitySource?: string;
  };
};

export const addGateway = createResourcePack(
  ({
    id,
    hostedZoneId,
    domainName,
    certificateArn,
    cloudFunction: { id: cloudFunctionId, region: cloudFunctionRegion = '${AWS::Region}' },
    stageName = 'production',
    authorizer,
    authorizer: {
      scopes: authScopes = ['phone', 'email', 'openid', 'profile'],
      type: authType = 'COGNITO_USER_POOLS',
      providerARNs,
      identitySource = 'method.request.header.authorization',
    } = {},
  }: AddGatewayConfig) => {
    const cloudFunctionUri = {
      'Fn::Sub': `arn:aws:apigateway:${cloudFunctionRegion}:lambda:path/2015-03-31/functions/\${${cloudFunctionId}.Arn}/invocations`,
    };
    const authorizerId = `${id}CustomAuthorizer`;
    const authProps = !!authorizer
      ? {
          AuthorizationScopes: authScopes,
          AuthorizationType: authType === DEFAULT_AUTH_TYPE ? DEFAULT_AUTH_TYPE : 'CUSTOM',
          AuthorizerId: {
            Ref: authorizerId,
          },
        }
      : undefined;

    return new SimpleCFT()
      .patch({
        Resources: {
          // REST API
          [`${id}GatewayRESTAPI`]: {
            Type: 'AWS::ApiGateway::RestApi',

            Properties: {
              Name: {
                'Fn::Sub': `\${AWS::StackName}-${id}GatewayRESTAPI`,
              },
              EndpointConfiguration: {
                Types: ['EDGE'],
              },
            },
          },
          [`${id}GatewayRESTAPIResource`]: {
            Type: 'AWS::ApiGateway::Resource',
            DependsOn: `${id}GatewayRESTAPI`,
            Properties: {
              ParentId: {
                'Fn::GetAtt': [`${id}GatewayRESTAPI`, 'RootResourceId'],
              },
              PathPart: '{proxy+}',
              RestApiId: {
                Ref: `${id}GatewayRESTAPI`,
              },
            },
          },
          [`${id}GatewayRESTAPIMethod`]: {
            Type: 'AWS::ApiGateway::Method',
            DependsOn: `${id}GatewayRESTAPIResource`,
            Properties: {
              ...authProps,
              HttpMethod: 'ANY',
              ResourceId: {
                Ref: `${id}GatewayRESTAPIResource`,
              },
              RestApiId: {
                Ref: `${id}GatewayRESTAPI`,
              },
              Integration: {
                Type: 'AWS_PROXY',
                IntegrationHttpMethod: 'POST',
                Uri: cloudFunctionUri,
              },
            },
          },
          [`${id}GatewayRESTAPIRootMethod`]: {
            Type: 'AWS::ApiGateway::Method',
            DependsOn: `${id}GatewayRESTAPIResource`,
            Properties: {
              ...authProps,
              HttpMethod: 'ANY',
              ResourceId: {
                'Fn::GetAtt': [`${id}GatewayRESTAPI`, 'RootResourceId'],
              },
              RestApiId: {
                Ref: `${id}GatewayRESTAPI`,
              },
              Integration: {
                Type: 'AWS_PROXY',
                IntegrationHttpMethod: 'POST',
                Uri: cloudFunctionUri,
              },
            },
          },
        },
      })
      .patch({
        Resources: {
          // CORS
          [`${id}GatewayRESTAPIOPTIONSMethod`]: {
            Type: 'AWS::ApiGateway::Method',
            DependsOn: `${id}GatewayRESTAPIResource`,
            Properties: {
              AuthorizationType: 'NONE',
              HttpMethod: 'OPTIONS',
              ResourceId: {
                Ref: `${id}GatewayRESTAPIResource`,
              },
              RestApiId: {
                Ref: `${id}GatewayRESTAPI`,
              },
              Integration: {
                Type: 'AWS_PROXY',
                IntegrationHttpMethod: 'POST',
                Uri: cloudFunctionUri,
              },
            },
          },
          [`${id}GatewayRESTAPIRootOPTIONSMethod`]: {
            Type: 'AWS::ApiGateway::Method',
            DependsOn: `${id}GatewayRESTAPIResource`,
            Properties: {
              AuthorizationType: 'NONE',
              HttpMethod: 'OPTIONS',
              ResourceId: {
                'Fn::GetAtt': [`${id}GatewayRESTAPI`, 'RootResourceId'],
              },
              RestApiId: {
                Ref: `${id}GatewayRESTAPI`,
              },
              Integration: {
                Type: 'AWS_PROXY',
                IntegrationHttpMethod: 'POST',
                Uri: cloudFunctionUri,
              },
            },
          },
          [`${id}GatewayResponseDefault4XX`]: {
            Type: 'AWS::ApiGateway::GatewayResponse',
            Properties: {
              ResponseParameters: {
                // Not authorized, so just allow the current origin by mapping it into the header.
                'gatewayresponse.header.Access-Control-Allow-Origin': 'method.request.header.origin',
                'gatewayresponse.header.Access-Control-Allow-Credentials': "'true'",
                'gatewayresponse.header.Access-Control-Allow-Headers': "'*'",
              },
              ResponseType: 'DEFAULT_4XX',
              RestApiId: {
                Ref: `${id}GatewayRESTAPI`,
              },
            },
          },
        },
      })
      .patch({
        Resources: {
          // SUPPORTING RESOURCES
          [`${id}GatewayRESTAPIDeployment`]: {
            Type: 'AWS::ApiGateway::Deployment',
            DependsOn: [
              `${id}GatewayRESTAPIResource`,
              `${id}GatewayRESTAPIMethod`,
              `${id}GatewayRESTAPIRootMethod`,
              `${id}GatewayRESTAPI`,
              cloudFunctionId,
            ],
            Properties: {
              RestApiId: {
                Ref: `${id}GatewayRESTAPI`,
              },
            },
          },
          [`${id}CloudWatch`]: {
            Type: 'AWS::Logs::LogGroup',
            Properties: {
              LogGroupName: {
                'Fn::Sub': `\${AWS::StackName}-${id}GatewayLogs`,
              },
            },
          },
          [`${id}CloudWatchRole`]: {
            Type: 'AWS::IAM::Role',
            Properties: {
              AssumeRolePolicyDocument: {
                Version: '2012-10-17T00:00:00.000Z',
                Statement: [
                  {
                    Effect: 'Allow',
                    Principal: {
                      Service: ['apigateway.amazonaws.com'],
                    },
                    Action: 'sts:AssumeRole',
                  },
                ],
              },
              Path: '/',
              ManagedPolicyArns: ['arn:aws:iam::aws:policy/service-role/AmazonAPIGatewayPushToCloudWatchLogs'],
            },
          },
          [`${id}CloudWatchAccount`]: {
            Type: 'AWS::ApiGateway::Account',
            Properties: {
              CloudWatchRoleArn: {
                'Fn::GetAtt': [`${id}CloudWatchRole`, 'Arn'],
              },
            },
          },
          [`${id}GatewayRESTAPIEnvironment`]: {
            Type: 'AWS::ApiGateway::Stage',
            DependsOn: ['APICloudWatchAccount', 'APIGatewayRESTAPIDeployment'],
            Properties: {
              AccessLogSetting: {
                DestinationArn: {
                  'Fn::GetAtt': [`${id}CloudWatch`, 'Arn'],
                },
                Format:
                  '{"requestId":"$context.requestId","ip":"$context.identity.sourceIp","caller":"$context.identity.caller","user":"$context.identity.user","requestTime":"$context.requestTime","httpMethod":"$context.httpMethod","resourcePath":"$context.resourcePath","status":"$context.status","protocol":"$context.protocol","responseLength":"$context.responseLength","apiGatewayErrorMessage":"$context.error.message"}',
              },
              DeploymentId: {
                Ref: 'APIGatewayRESTAPIDeployment',
              },
              RestApiId: {
                Ref: 'APIGatewayRESTAPI',
              },
              StageName: stageName,
            },
          },
        },
      })
      .patch({
        Resources: {
          // DNS
          [`${id}DomainName`]: {
            Type: 'AWS::ApiGateway::DomainName',
            Properties: {
              CertificateArn: certificateArn,
              DomainName: domainName,
              EndpointConfiguration: {
                Types: ['EDGE'],
              },
            },
          },
          [`${id}DomainNameBasePathMapping`]: {
            Type: 'AWS::ApiGateway::BasePathMapping',
            DependsOn: [`${id}GatewayRESTAPI`, `${id}GatewayRESTAPIEnvironment`, `${id}DomainName`],
            Properties: {
              DomainName: domainName,
              RestApiId: {
                Ref: `${id}GatewayRESTAPI`,
              },
              Stage: stageName,
            },
          },
          [`${id}Route53Record`]: {
            Type: 'AWS::Route53::RecordSet',
            DependsOn: `${id}DomainName`,
            Properties: {
              HostedZoneId: hostedZoneId,
              Type: 'A',
              Name: {
                'Fn::Sub': [
                  '${DomainName}.',
                  {
                    DomainName: domainName,
                  },
                ],
              },
              AliasTarget: {
                HostedZoneId: 'Z2FDTNDATAQYW2',
                DNSName: {
                  'Fn::Sub': [
                    '${DomainName}.',
                    {
                      DomainName: {
                        'Fn::GetAtt': [`${id}DomainName`, 'DistributionDomainName'],
                      },
                    },
                  ],
                },
              },
            },
          },
        },
      })
      .patch({
        Resources: {
          // PERMISSIONS
          [`${id}CloudFunctionANYResourcePermission`]: {
            Type: 'AWS::Lambda::Permission',
            Properties: {
              Action: 'lambda:InvokeFunction',
              Principal: 'apigateway.amazonaws.com',
              FunctionName: {
                Ref: 'APICloudFunction',
              },
              SourceArn: {
                'Fn::Sub': [
                  'arn:aws:execute-api:${AWS::Region}:${AWS::AccountId}:${__ApiId__}/${__Stage__}/*/*',
                  {
                    // TODO: Fix!
                    __Stage__: '',
                    __ApiId__: {
                      Ref: 'APIGatewayRESTAPI',
                    },
                  },
                ],
              },
            },
          },
        },
      })
      .patch(
        !!authorizer
          ? {
              Resources: {
                // AUTHORIZER
                [`${id}CustomAuthorizer`]: {
                  Type: 'AWS::ApiGateway::Authorizer',
                  Properties: {
                    IdentitySource: identitySource,
                    Name: `${id}CustomAuthorizer`,
                    ProviderARNs: providerARNs,
                    RestApiId: {
                      Ref: `${id}GatewayRESTAPI`,
                    },
                    Type: 'COGNITO_USER_POOLS',
                  },
                },
              },
            }
          : {}
      ).template;
  }
);
