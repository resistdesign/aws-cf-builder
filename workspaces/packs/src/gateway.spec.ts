import { addGateway } from './gateway';

describe('Gateway', () => {
  describe('addGateway', () => {
    test('should add a gateway for an API', () => {
      const cft = addGateway(
        {
          id: 'Basic',
          hostedZoneId: '1234567890',
          domainName: 'api.example.com',
          certificateArn: 'i:am:a:cert:arn',
          stageName: 'production',
          cloudFunction: {
            id: 'CFID',
            region: 'us-east-1',
          },
          authorizer: {
            providerARNs: ['PROVIDER_ARN'],
          },
        },
        { AWSTemplateFormatVersion: '2010-09-09' }
      );

      expect(cft).toMatchSnapshot();
    });
  });
});
