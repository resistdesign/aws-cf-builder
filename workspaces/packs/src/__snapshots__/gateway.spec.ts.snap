// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gateway addGateway should add a gateway for an API 1`] = `
Object {
  "AWSTemplateFormatVersion": "2010-09-09",
  "Resources": Object {
    "BasicCloudFunctionANYResourcePermission": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Ref": "BasicCloudFunction",
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Sub": Array [
            "arn:aws:execute-api:\${AWS::Region}:\${AWS::AccountId}:\${__ApiId__}/\${__Stage__}/*/*",
            Object {
              "__ApiId__": Object {
                "Ref": "BasicGatewayRESTAPI",
              },
              "__Stage__": "production",
            },
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "BasicCloudWatch": Object {
      "Properties": Object {
        "LogGroupName": Object {
          "Fn::Sub": "\${AWS::StackName}-BasicGatewayLogs",
        },
      },
      "Type": "AWS::Logs::LogGroup",
    },
    "BasicCloudWatchAccount": Object {
      "Properties": Object {
        "CloudWatchRoleArn": Object {
          "Fn::GetAtt": Array [
            "BasicCloudWatchRole",
            "Arn",
          ],
        },
      },
      "Type": "AWS::ApiGateway::Account",
    },
    "BasicCloudWatchRole": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": Array [
                  "apigateway.amazonaws.com",
                ],
              },
            },
          ],
          "Version": "2012-10-17T00:00:00.000Z",
        },
        "ManagedPolicyArns": Array [
          "arn:aws:iam::aws:policy/service-role/AmazonAPIGatewayPushToCloudWatchLogs",
        ],
        "Path": "/",
      },
      "Type": "AWS::IAM::Role",
    },
    "BasicCustomAuthorizer": Object {
      "Properties": Object {
        "IdentitySource": "method.request.header.authorization",
        "Name": "BasicCustomAuthorizer",
        "ProviderARNs": Array [
          "PROVIDER_ARN",
        ],
        "RestApiId": Object {
          "Ref": "BasicGatewayRESTAPI",
        },
        "Type": "COGNITO_USER_POOLS",
      },
      "Type": "AWS::ApiGateway::Authorizer",
    },
    "BasicDomainName": Object {
      "Properties": Object {
        "CertificateArn": "i:am:a:cert:arn",
        "DomainName": "api.example.com",
        "EndpointConfiguration": Object {
          "Types": Array [
            "EDGE",
          ],
        },
      },
      "Type": "AWS::ApiGateway::DomainName",
    },
    "BasicDomainNameBasePathMapping": Object {
      "DependsOn": Array [
        "BasicGatewayRESTAPI",
        "BasicGatewayRESTAPIEnvironment",
        "BasicDomainName",
      ],
      "Properties": Object {
        "DomainName": "api.example.com",
        "RestApiId": Object {
          "Ref": "BasicGatewayRESTAPI",
        },
        "Stage": "production",
      },
      "Type": "AWS::ApiGateway::BasePathMapping",
    },
    "BasicGatewayRESTAPI": Object {
      "Properties": Object {
        "EndpointConfiguration": Object {
          "Types": Array [
            "EDGE",
          ],
        },
        "Name": Object {
          "Fn::Sub": "\${AWS::StackName}-BasicGatewayRESTAPI",
        },
      },
      "Type": "AWS::ApiGateway::RestApi",
    },
    "BasicGatewayRESTAPIDeployment": Object {
      "DependsOn": Array [
        "BasicGatewayRESTAPIResource",
        "BasicGatewayRESTAPIMethod",
        "BasicGatewayRESTAPIRootMethod",
        "BasicGatewayRESTAPI",
        "CFID",
      ],
      "Properties": Object {
        "RestApiId": Object {
          "Ref": "BasicGatewayRESTAPI",
        },
      },
      "Type": "AWS::ApiGateway::Deployment",
    },
    "BasicGatewayRESTAPIEnvironment": Object {
      "DependsOn": Array [
        "APICloudWatchAccount",
        "APIGatewayRESTAPIDeployment",
      ],
      "Properties": Object {
        "AccessLogSetting": Object {
          "DestinationArn": Object {
            "Fn::GetAtt": Array [
              "BasicCloudWatch",
              "Arn",
            ],
          },
          "Format": "{\\"requestId\\":\\"$context.requestId\\",\\"ip\\":\\"$context.identity.sourceIp\\",\\"caller\\":\\"$context.identity.caller\\",\\"user\\":\\"$context.identity.user\\",\\"requestTime\\":\\"$context.requestTime\\",\\"httpMethod\\":\\"$context.httpMethod\\",\\"resourcePath\\":\\"$context.resourcePath\\",\\"status\\":\\"$context.status\\",\\"protocol\\":\\"$context.protocol\\",\\"responseLength\\":\\"$context.responseLength\\",\\"apiGatewayErrorMessage\\":\\"$context.error.message\\"}",
        },
        "DeploymentId": Object {
          "Ref": "APIGatewayRESTAPIDeployment",
        },
        "RestApiId": Object {
          "Ref": "APIGatewayRESTAPI",
        },
        "StageName": "production",
      },
      "Type": "AWS::ApiGateway::Stage",
    },
    "BasicGatewayRESTAPIMethod": Object {
      "DependsOn": "BasicGatewayRESTAPIResource",
      "Properties": Object {
        "AuthorizationScopes": Array [
          "phone",
          "email",
          "openid",
          "profile",
        ],
        "AuthorizationType": "COGNITO_USER_POOLS",
        "AuthorizerId": Object {
          "Ref": "BasicCustomAuthorizer",
        },
        "HttpMethod": "ANY",
        "Integration": Object {
          "IntegrationHttpMethod": "POST",
          "Type": "AWS_PROXY",
          "Uri": Object {
            "Fn::Sub": "arn:aws:apigateway:us-east-1:lambda:path/2015-03-31/functions/\${CFID.Arn}/invocations",
          },
        },
        "ResourceId": Object {
          "Ref": "BasicGatewayRESTAPIResource",
        },
        "RestApiId": Object {
          "Ref": "BasicGatewayRESTAPI",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
    "BasicGatewayRESTAPIOPTIONSMethod": Object {
      "DependsOn": "BasicGatewayRESTAPIResource",
      "Properties": Object {
        "AuthorizationType": "NONE",
        "HttpMethod": "OPTIONS",
        "Integration": Object {
          "IntegrationHttpMethod": "POST",
          "Type": "AWS_PROXY",
          "Uri": Object {
            "Fn::Sub": "arn:aws:apigateway:us-east-1:lambda:path/2015-03-31/functions/\${CFID.Arn}/invocations",
          },
        },
        "ResourceId": Object {
          "Ref": "BasicGatewayRESTAPIResource",
        },
        "RestApiId": Object {
          "Ref": "BasicGatewayRESTAPI",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
    "BasicGatewayRESTAPIResource": Object {
      "DependsOn": "BasicGatewayRESTAPI",
      "Properties": Object {
        "ParentId": Object {
          "Fn::GetAtt": Array [
            "BasicGatewayRESTAPI",
            "RootResourceId",
          ],
        },
        "PathPart": "{proxy+}",
        "RestApiId": Object {
          "Ref": "BasicGatewayRESTAPI",
        },
      },
      "Type": "AWS::ApiGateway::Resource",
    },
    "BasicGatewayRESTAPIRootMethod": Object {
      "DependsOn": "BasicGatewayRESTAPIResource",
      "Properties": Object {
        "AuthorizationScopes": Array [
          "phone",
          "email",
          "openid",
          "profile",
        ],
        "AuthorizationType": "COGNITO_USER_POOLS",
        "AuthorizerId": Object {
          "Ref": "BasicCustomAuthorizer",
        },
        "HttpMethod": "ANY",
        "Integration": Object {
          "IntegrationHttpMethod": "POST",
          "Type": "AWS_PROXY",
          "Uri": Object {
            "Fn::Sub": "arn:aws:apigateway:us-east-1:lambda:path/2015-03-31/functions/\${CFID.Arn}/invocations",
          },
        },
        "ResourceId": Object {
          "Fn::GetAtt": Array [
            "BasicGatewayRESTAPI",
            "RootResourceId",
          ],
        },
        "RestApiId": Object {
          "Ref": "BasicGatewayRESTAPI",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
    "BasicGatewayRESTAPIRootOPTIONSMethod": Object {
      "DependsOn": "BasicGatewayRESTAPIResource",
      "Properties": Object {
        "AuthorizationType": "NONE",
        "HttpMethod": "OPTIONS",
        "Integration": Object {
          "IntegrationHttpMethod": "POST",
          "Type": "AWS_PROXY",
          "Uri": Object {
            "Fn::Sub": "arn:aws:apigateway:us-east-1:lambda:path/2015-03-31/functions/\${CFID.Arn}/invocations",
          },
        },
        "ResourceId": Object {
          "Fn::GetAtt": Array [
            "BasicGatewayRESTAPI",
            "RootResourceId",
          ],
        },
        "RestApiId": Object {
          "Ref": "BasicGatewayRESTAPI",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
    "BasicGatewayResponseDefault4XX": Object {
      "Properties": Object {
        "ResponseParameters": Object {
          "gatewayresponse.header.Access-Control-Allow-Credentials": "'true'",
          "gatewayresponse.header.Access-Control-Allow-Headers": "'*'",
          "gatewayresponse.header.Access-Control-Allow-Origin": "method.request.header.origin",
        },
        "ResponseType": "DEFAULT_4XX",
        "RestApiId": Object {
          "Ref": "BasicGatewayRESTAPI",
        },
      },
      "Type": "AWS::ApiGateway::GatewayResponse",
    },
    "BasicRoute53Record": Object {
      "DependsOn": "BasicDomainName",
      "Properties": Object {
        "AliasTarget": Object {
          "DNSName": Object {
            "Fn::Sub": Array [
              "\${DomainName}.",
              Object {
                "DomainName": Object {
                  "Fn::GetAtt": Array [
                    "BasicDomainName",
                    "DistributionDomainName",
                  ],
                },
              },
            ],
          },
          "HostedZoneId": "Z2FDTNDATAQYW2",
        },
        "HostedZoneId": "1234567890",
        "Name": Object {
          "Fn::Sub": Array [
            "\${DomainName}.",
            Object {
              "DomainName": "api.example.com",
            },
          ],
        },
        "Type": "A",
      },
      "Type": "AWS::Route53::RecordSet",
    },
  },
}
`;
